geth/peers/peerpool.go

```diff
@@ -0,0 +1,278 @@
+	fast := true
+	period <- p.fastSync
+	connected := 0
+	selfID := discv5.NodeID(server.Self().ID)
+	for {
+		select {
+		case <-p.quit:
+			return
+		case node := <-found:
+			log.Debug("found node with", "ID", node.ID, "topic", topic)
+			if node.ID == selfID {
+				continue
+			}
+			if p.processFoundNode(server, connected, topic, node) {
+				connected++
+			}
+			// switch period only once
+			if fast && connected >= limits[0] {
+				log.Debug("switch to slow sync", "topic", topic, "sync", p.slowSync)
+				period <- p.slowSync
+				fast = false
+			}
+		case <-lookup:
+			// just drain this channel for now, it can be used to intelligently
+			// limit number of kademlia lookups
+		case event := <-events:
+			if event.Type == p2p.PeerEventTypeDrop {
+				log.Debug("node dropped", "ID", event.Peer, "topic", topic)
+				if !p.processDroppedNode(server, topic, event.Peer) {
+					connected--
+				}
+				// switch period only once
+				if !fast && connected < limits[0] {
+					log.Debug("switch to fast sync", "topic", topic, "sync", p.fastSync)
+					period <- p.fastSync
+					fast = true
+				}
+			}
+		}
+	}
+}
+
+// processFoundNode called when node is discovered by kademlia search query
+// 2 important conditions
+// 1. every time when node is processed we need to update discoveredTime and reset dropped boolean.
+//    peer will be considered as valid later only if it was discovered < 90s ago and wasn't dropped recently
+// 2. if peer is connected or if max limit is reached we are not a adding peer to p2p server
+func (p *PeerPool) processFoundNode(server *p2p.Server, currentlyConnected int, topic discv5.Topic, node *discv5.Node) (connected bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	peersTable := p.peers[topic]
+	limits := p.config[topic]
+	if info, exist := peersTable[node.ID]; exist {
+		info.discoveredTime = mclock.Now()
+	} else {
+		peersTable[node.ID] = &peerInfo{
+			discoveredTime: mclock.Now(),
+			node:           node,
+		}
+	}
+	if currentlyConnected < limits[1] && !peersTable[node.ID].connected {
+		log.Debug("peer connected", "ID", node.ID, "topic", topic)
+		server.AddPeer(discover.NewNode(
+			discover.NodeID(peersTable[node.ID].node.ID),
+			peersTable[node.ID].node.IP,
+			peersTable[node.ID].node.UDP,
+			peersTable[node.ID].node.TCP,
+		))
+		peersTable[node.ID].connected = true
+		connected = true
+		if p.cache != nil {
+			if err := p.cache.AddPeer(node, topic); err != nil {
+				log.Error("failed to persist a peer", "error", err)
+			}
+		}
+	}
+	return connected
+}
+
+// processDroppedNode is called when node was dropped by p2p server
+// - removes a peer, cause p2p server now relies on peer pool to maintain required connections number
+// - if there is a valid peer in peer table add it to a p2p server
+//   peer is valid if it wasn't dropped recently, it is not stale (90s) and not currently connected
+func (p *PeerPool) processDroppedNode(server *p2p.Server, topic discv5.Topic, nodeID discover.NodeID) bool {
```

Comment:
I would add a comment about the return value. Why it's bool and what `true` means.